numpy_ml/cluster/kmeans.py

```diff
@@ -0,0 +1,210 @@
+
+                    if dist < min_distance:                
+                        min_distance = dist
+                        assignments[i] = k
+
+            # update centroids
+            set_labels = range(self.n_clusters)
+            # filter by labels
+            for label in set_labels:
+                filter_indices = np.where(np.array(assignments) == label)[0]
+                count_per_label = len(filter_indices)
+                if count_per_label > 0:
+                    filter_xdata = np.take(self.X_train, filter_indices, axis=0)
+                    centroids[label, :] = np.mean(filter_xdata, axis=0)
+            weight_list.append(centroids)
+            weight_list.pop(0)
+            iteration = iteration + 1
+        return centroids, assignments
+
+
+    def _kmeans_soft(self):
+        r"""
+        Soft clustering: In this implementation, which is the modification of the kmean hard  clustering algorithm, we assigns every data point a degree of membership in the cluster assignment. Hence, we give a probability distribution.
+
+        Parameters:
+        -----------
+        beta: float
+            span of a radial basis kernel
+
+        Returns
+        -------
+        centroids : array
+                    The centroid matrix with dimensions (number of centroids x number of attributes).
+        assignments : array
+                    The assignments vector with dimensions (number of data x number of centroids). This is the probability distribution of the membership of each data point in the cluster.
+        """
+        size_of_data = self.X_train.shape[0]
+        centroid_indexes = np.random.choice(size_of_data, self.n_clusters, replace=False)
+        centroids = np.take(self.X_train, centroid_indexes, axis=0)
+        assignments = -1 * np.ones((size_of_data, self.n_clusters))
+        n_dims = self.X_train.shape[1]
+        iteration, max_iteration = 0, 100 
+        tol = 0.00001 # prevent division by zero
+        prev_weight_vec = np.zeros(n_dims)
+        weight_list = [prev_weight_vec, centroids]  
+
+        while not self._is_converged(weight_list[-2], weight_list[-1]) or iteration < max_iteration:
+            centroids = weight_list[-1]
+
+            # update cluster assignments
+            for i, x_val in enumerate(self.X_train):
+                for k, mu_val in enumerate(centroids):
+                    dist = np.linalg.norm(x_val - mu_val)
+                    weight = np.exp(-dist / self.beta)
+                    assignments[i][k] = weight
+            # normalize assignment matrix
+            row_sums = assignments.sum(axis=1)
+            assignments = (assignments + tol) / (row_sums[:, np.newaxis] + tol)
+
+            # update centroids
+            set_labels = range(self.n_clusters)
+            # filter by labels
+            for label in set_labels:
+                filter_indices = np.where(np.array(assignments) == label)[0]
```

Comment:
this doesn't seem right, since your assignment is float by "soft" target, and label is int...

Reply:
Thanks for your feedback @qianyizhang. However, in this variation of kmeans algorithm, we are assigning degree of memberships in the label assignment. It is not a hard assignment as in the hard version of kmeans algorithm

Reply:
to be more specific, np.array(assignments) is bunch of **floats**, and label is **int**.
They are not going to be equal, and filter_indices  is therefore broken

Reply:
Thanks so much for the catch. I will update my implementation. I made a mistake. I will follow https://user-images.githubusercontent.com/1905599/119421211-0bea3b80-bcb3-11eb-9e71-a337da8db24d.png more closely in my next commit to the branch

Reply:
@qianyizhang  think that I need something like this to update centroid, but I have not yet tested for dimensions mismatch. I will test and get the correct logic this weekend
```
            # update centroids
            for x_val in enumerate(self.X_train):
                for label in set_labels:
                    weight_by_label = np.sum(x_val * assignments[:,label])

                    # estimate centroid
                    numerator = weight_by_label # sum the weighted inputs
                    denominator = np.sum(weight_by_label) # weight contribution across data input

                    centroids[label, :] = (numerator + tol) / (denominator + tol)
```

Reply:
@qianyizhang This is the right logic
```
            # update centroids
            set_labels = range(self.n_clusters)
            for label in set_labels:
                numerator = np.zeros((1, n_dims))
                denominator = np.zeros((1, n_dims))
                for x_val in enumerate(self.X_train):
                    # estimate centroid
                    numerator += x_val * assignments[:,label] # weight contribution across data input
                    denominator += assignments[:, label] 
                centroids[label, :] = (numerator + tol) / (denominator + tol)

            weight_list.append(centroids)
```

Let me test more before updating the PR

Reply:
@ddbourgin @qianyizhang I have fixed the bug. Please take another look at the PR